client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+# Treode Client Library Interface
+
+The two major components of the Treode client library interface are the [Cache](#Cache) and the [Transaction](#Transaction).  
+
+Caches store key-value pairs retrieved from the server, as well as the read and value timestamps associated with each pair that Treode provides.  Caches are responsible for maintaining a connection to the DB server.  Users may create multiple caches to work on disjoint sections of the database simultaneously.  
+
+Transactions allows users to locally perform a sequence of database operations and, when finished, optimistically commit the result to the server by writing through the cache.  Users can choose upon creating a Transaction which cache to associate it with.
+
+<a name="Cache"></a>
+# Cache
+
+We provide an overview of the Cache class below, including its primary internal data structure, its public methods, and the form of its HTTP requests and responses.
+
+- Parameters: 
+    1. server: string (required)
+    2. port: string (port=80)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, read\_timestamp=currentTime, max\_age=None, no\_cache=False): value (JSON)
+    2. write(table: string, op\_list: list): status (string), throws StaleException
+- Internal Objects: 
+    1. HTTP library for establishing/managing connection to server (in Python, urllib3)
+    2. sorted map for maintaining cache (in Python, [skiplist](https://github.com/sepeth/python-skiplist))
```

Comment:
This skip list doesn't have floor and ceiling. We need floor and ceiling for our lookups.